### ISSUE TITLE
Timestamp precision of events is low

### ISSUE THREAD

**Issue by senecobis** [better-flow/evimo#32]
Hi @aftersomemath and @ncos thanks for you great work.

I raised this issue as both in EVIMO and EVIMO2 I noticed that the precision of the timestamps of the events are around 10^-2.
This effectively transform the stream of events in a stream of sparse frames.
I looked inside the rosbags, txts and npzs of EVIMO and EVIMO2 but seems that this problem is commonly spread.

I guess this issue is connected with [this](https://github.com/better-flow/evimo/issues/20) however my question is, are there specific trajectories that didn't suffer from this problem ? If not do you plan to release a corrected version ?

**Comment by aftersomemath** [comment]
Hello!

If you provide a short example reproducing the issue you are facing I will be happy to take a look (in both datasets). But at least in EVIMO2, the timestamps have a precision of about 10^-4 seconds for the samsung camera and about 10^-6 seconds for the prophesee cameras. This precision is determined by the camera hardware's time resolution.

Here are plots of the gradient of the event timestamps for each camera in the `scene13_dyn_test_00_000000` imo sequence. The maximum time delta between two events is about 10^-4 for the samsung camera and 2*10^-5 for the prophesee cameras (though often on the order of 10^-6). The same conclusion can be drawn from the plots in the older issue referenced (#20) which were generated using the entire dataset.

![Figure_1](https://github.com/user-attachments/assets/d401514e-0557-498a-bc42-f12ebc9216c6)

Here is the code to produce the plot above:
```python
import numpy as np
import matplotlib.pyplot as plt

t_samsung = np.load('samsung_mono/imo/eval/scene13_dyn_test_00_000000/dataset_events_t.npy')
t_left = np.load('left_camera/imo/eval/scene13_dyn_test_00_000000/dataset_events_t.npy')
t_right = np.load('right_camera/imo/eval/scene13_dyn_test_00_000000/dataset_events_t.npy')

plt.subplot(311)
plt.plot(t_samsung, np.gradient(t_samsung))
plt.title('samsung mono')
plt.subplot(312)
plt.plot(t_left, np.gradient(t_left))
plt.title('prophesee left')
plt.subplot(313)
plt.plot(t_right, np.gradient(t_right))
plt.title('prophesee right')

plt.tight_layout()
plt.show()
```

**Comment by senecobis** [comment]
Hi.
Basically I download the data from the [script](https://github.com/better-flow/evimo/tree/master/tools/evimo2_v2_downloader).
Then I use the following function to import the data:

```
def load_samsung(file):
    meta  = np.load(os.path.join(file, 'dataset_info.npz'), allow_pickle=True)['meta'].item()
    depth = np.load(os.path.join(file, 'dataset_depth.npz'))
    mask  = np.load(os.path.join(file, 'dataset_mask.npz'))
    classical = np.load(os.path.join(file, 'dataset_classical.npz'))
    events_xy = np.load(os.path.join(file, 'dataset_events_xy.npy'))
    events_p = np.load(os.path.join(file, 'dataset_events_p.npy'))
    events_t = np.load(os.path.join(file, 'dataset_events_t.npy'), allow_pickle=True).astype(np.float32)
    data = {
        'meta': meta,
        'depth': depth,
        'mask': mask,
        'classical': classical,
        'events_xy': events_xy,
        'events_p': events_p,
        'events_t': events_t
    }
    return data
```
    
For the second sequence for instance and for the first 10 events from left_cam I get:

`array([2.25    , 2.25    , 2.25    , 2.25    , 2.25    , 2.25    , 2.25    , 2.25    , 2.250001, 2.250001], dtype=float32)`

Instead from samsung I get:

`array([2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012], dtype=float32)`

For samsung for instance the first 1000 events have all the same tstamps:

`array([2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012,
       2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012,
       2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012,
       ...
       2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012,
       2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012,
       2.25012, 2.25012, 2.25012, 2.25012, 2.25012, 2.25012],
      dtype=float32)`

For the first sequence the distribution of timestamps is a little better but still significantly discrete.
The fist 50 events from samsung are:

`array([0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01686268, 0.01686268,
       0.01686268, 0.01686268, 0.01686268, 0.01706773, 0.01706773,
       0.01706773, 0.01706773, 0.01706773, 0.01706773, 0.01706773,
       0.01706773, 0.01706773, 0.01706773, 0.01706773, 0.01706773],
      dtype=float32)`
      

Is this normal or am I missing something ?

**Comment by aftersomemath** [comment]
Yes this is normal and consistent with a precision of 10^-6 for the prophesee cameras (left, right) and 10^-4 for the samsung camera as supported by the plot in my first reply and those in #20.

Events will be grouped together according to the event rate and the time resolution of the camera. For example if the instantaneous event rate is, for example, 6 million events/s and the time resolution is 1e-6, about 6 events will have the same timestamp. At 1e-4 time precision, 600 events will have the same timestamp. Of course, the rates will not be uniform, but this gives a sense of the order of magnitude of the events in each group for each camera.

In the first example with the prophesee camera, after 8 events the timestamps increment by 10^-6. This indicates that the time precision is 10^-6. Additionally, this corresponds with an event rate of 8 million events per second, which is reasonable.

In the second example, the Samsung camera's events have the same timestamp for at least the first 1000 events. While we can't conclude anything about the time precision from this, if the precision is 10^-4, then this suggests an event rate of at least 10 million events per second, which is reasonable.

In the third example, the Samsung camera's timestamps increment by 0.01706773 - 0.01686268~=2*10^-4 after ~40 events. The low number of events and the skipped time increment (assuming time precision of 10^-4) is most likely because the camera is not generating many events at the beginning of the recording (as can be seen in [the video](https://obj.umiacs.umd.edu/evimo2v2videowebsite/imo_scene13_dyn_test_00_000000.mp4) for this sequence). It is also possible some events with the same timestamp as the first event in this sequence were discarded before the recording started.

A visual way that the time precision is much better than 10^-2 is possible with the [viewer tool](https://better-flow.github.io/evimo/docs/evimo2v2-inspect-sequence.html). It has slider that allows visualizing event slices with 1 millisecond precision. The visualized events move smoothly as the boundaries of the slice are moved in 1 millisecond increments.
